### PR TITLE
feat(*) retry policy

### PR DESCRIPTION
**Pull request by bartsmykla**
### Summary

Introducing retry policy with option to configure http/grpc/tcp policies

GUI changes: https://github.com/kumahq/kuma-gui/pull/108/

### Full changelog

TBD

### Issues resolved
None

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/XXX)

### CODE REVIEW

**Comment by subnetmarco** [comment]
+ GUI

**Review by zawiszaty — APPROVED** [review]
lgfm :sunglasses:

File: api/mesh/v1alpha1/retry.proto [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+      //  +optional
+      google.protobuf.Duration per_try_timeout = 3;
+
+      //  +optional
+      BackOff back_off = 4;
+    }
+
+    //  +required
+    oneof protocol {
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/retry.proto:72]: I think `oneof` may not be a good idea. Imagine we want to add default retry policy. We don't know which protocols are used, but we want to retry on connection failures. We need to create 3 default policies then, one for tcp, one for HTTP and one for grpc. But the problem is that we can match only one policy per outbound and we don't have information about protocol so effectively only one (latest) will be chosen.

**Review by jakubdyszkiewicz — COMMENTED** [review]
Very solid PR. A couple of nits here and there.

File: pkg/xds/generator/outbound_proxy_generator.go [hunk 2]
```diff
@@ -90,25 +91,51 @@ func (_ OutboundProxyGenerator) generateLDS(proxy *model.Proxy, subsets []envoy_
...
-				Configure(envoy_listeners.HttpAccessLog(meshName, envoy_common.TrafficDirectionOutbound, sourceService, serviceName, proxy.Policies.Logs[serviceName], proxy)).
-				Configure(envoy_listeners.HttpOutboundRoute(serviceName, subsets, proxy.Dataplane.Spec.TagSet()))
+				Configure(envoy_listeners.HttpAccessLog(
+					meshName,
+					envoy_common.TrafficDirectionOutbound,
+					sourceService,
+					serviceName,
+					proxy.Policies.Logs[serviceName],
+					proxy,
+				)).
```
[the remark this answers is not in the document]
Comment (bartsmykla) [reply pkg/xds/generator/outbound_proxy_generator.go:107]: Agree, apologies, my IDE was probably involved when resolving conflicts

File: pkg/xds/envoy/listeners/v2/retry_configurer.go [hunk 3]
```diff
@@ -13,10 +13,12 @@ import (
...
-	HttpRetryOn5XX                  = "5xx"
-	HttpRetryOnRetriableStatusCodes = "retriable-status-codes"
-	GrpcRetryOnAll                  = "cancelled,deadline-exceeded,internal," +
-		"resource-exhausted,unavailable"
+	HttpRetryOnDefault = "gateway-error,connect-failure," +
+		"refused-stream"
+	HttpRetryOnRetriableStatusCodes = "connect-failure,refused-stream," +
+		"retriable-status-codes"
+	GrpcRetryOnAll = "cancelled,connect-failure," +
+		"gateway-error,refused-stream,reset,resource-exhausted,unavailable"
```
Comment (jakubdyszkiewicz) [inline pkg/xds/envoy/listeners/v2/retry_configurer.go:21]: would be great to add a comment on why we picked those default values

File: api/mesh/v1alpha1/retry.proto (hunk 1 above)
Comment (bartsmykla) [reply api/mesh/v1alpha1/retry.proto]: Addressed

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**